File: src/dk/eightyplus/Painter/fragment/ColorPaletteFragment.java
package dk.eightyplus.Painter.fragment;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.LinearInterpolator;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.Toast;
import dk.eightyplus.Painter.Callback;
import dk.eightyplus.Painter.Keys;
import dk.eightyplus.Painter.R;
import dk.eightyplus.Painter.dialog.ColorPickerDialog;
import dk.eightyplus.Painter.dialog.ColorPickerView;

import java.lang.ref.SoftReference;

/**
 * User: fries
 * Date: 15/07/14
 * Time: 14.00
 */
public class ColorPaletteFragment extends DialogFragment {

  private SoftReference<Callback> callbackSoftReference;
  private LinearLayout linearLayoutDynamic;

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    this.callbackSoftReference = new SoftReference<Callback>((Callback) getActivity());
  }

  @Override
  public View onCreateView(final LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    View view = inflater.inflate(R.layout.color_palette, container, false);

    Callback callback = callbackSoftReference.get();

    LinearLayout linearLayoutTop = (LinearLayout) view.findViewById(R.id.color_top);
    LinearLayout blackWhite = (LinearLayout) view.findViewById(R.id.color_static1);
    LinearLayout grayScale = (LinearLayout) view.findViewById(R.id.color_static2);
    LinearLayout rainbow = (LinearLayout) view.findViewById(R.id.color_rainbow);
    linearLayoutDynamic = (LinearLayout) view.findViewById(R.id.color_dynamic);

    int[] bwColors = new int[]{R.color.black, R.color.white};
    for (int color : bwColors) {
      final int hexColor = getActivity().getResources().getColor(color);

      ImageButton button = (ImageButton) inflater.inflate(R.layout.color_button, blackWhite, false);
      button.setBackgroundColor(hexColor);
      blackWhite.addView(button);

      button.setOnClickListener(new ButtonColorClickListener(callback, hexColor));
    }
    ImageButton randomButton = (ImageButton) inflater.inflate(R.layout.color_button, blackWhite, false);
    randomButton.setBackgroundResource(android.R.drawable.ic_menu_help);
    randomButton.setOnClickListener(new ButtonColorClickListener(callback, 0));
    blackWhite.addView(randomButton);

    int[] grayColors = new int[]{R.color.dark_grey, R.color.gray, R.color.light_grey};
    for (int color : grayColors) {
      final int hexColor = getActivity().getResources().getColor(color);

      ImageButton button = (ImageButton) inflater.inflate(R.layout.color_button, grayScale, false);
      button.setBackgroundColor(hexColor);
      grayScale.addView(button);

      button.setOnClickListener(new ButtonColorClickListener(callback, hexColor));
    }

    ColorPickerView colorPickerView = new ColorPickerView(inflater.getContext(), new ColorPickerDialog.OnColorChangedListener() {
      private int lastColor = 0;

      @Override
      public void colorChanged(int color) {
        Callback callback = callbackSoftReference.get();
        if (callback != null) {
          callback.colorChanged(color);
        }

        if (lastColor != color) {
          lastColor = color;

          final SharedPreferences preferences = getActivity().getSharedPreferences(Keys.PREFERENCES, Context.MODE_PRIVATE);
          int lastColorSlot = preferences.getInt(Keys.COLOR_LAST, -1);

          lastColorSlot++;
          if (lastColorSlot > 5) {
            lastColorSlot = 0;
          }

          SharedPreferences.Editor edit = preferences.edit();
          edit.putInt(Keys.COLOR + lastColorSlot, color);
          edit.putInt(Keys.COLOR_LAST, lastColorSlot);
          edit.commit();

          updatePickedColors(inflater, callback);
          linearLayoutDynamic.invalidate();
        }
      }
    }, 0);
    linearLayoutTop.addView(colorPickerView);

    int[] colors = new int[]{R.color.red, R.color.Orange, R.color.yellow, R.color.green, R.color.blue, R.color.purple};
    for (int color : colors) {
      final int hexColor = getActivity().getResources().getColor(color);

      ImageButton button = (ImageButton) inflater.inflate(R.layout.color_button, rainbow, false);
      button.setBackgroundColor(hexColor);
      rainbow.addView(button);

      button.setOnClickListener(new ButtonColorClickListener(callback, hexColor));
    }

    updatePickedColors(inflater, callback);

    return view;
  }

  private void updatePickedColors(LayoutInflater inflater, Callback callback) {
    linearLayoutDynamic.removeAllViews();
    final SharedPreferences preferences = getActivity().getSharedPreferences(Keys.PREFERENCES, Context.MODE_PRIVATE);
    //clearSavedColors(preferences);
    int lastColorSlot = preferences.getInt(Keys.COLOR_LAST, -1);
    if (lastColorSlot >= 0) {
      for (int i = 0 ; i < 6; i++) {
        final int hexWhiteColor = getActivity().getResources().getColor(R.color.white);
        int hexColor = preferences.getInt(Keys.COLOR + i, hexWhiteColor);

        ImageButton button = (ImageButton) inflater.inflate(R.layout.color_button, linearLayoutDynamic, false);
        button.setBackgroundColor(hexColor);
        linearLayoutDynamic.addView(button);

        button.setOnClickListener(new ButtonColorClickListener(callback, hexColor));
      }
    }
  }

  private void clearSavedColors(SharedPreferences preferences) {
    SharedPreferences.Editor editor = preferences.edit();
    editor.remove(Keys.COLOR_LAST);

    for (int i = 0 ; i < 6; i++) {
      editor.remove(Keys.COLOR + i);
    }
    editor.commit();
  }

  private static class ButtonColorClickListener implements View.OnClickListener {

    private SoftReference<Callback> callbackSoftReference;
    private int color;
    private Animation animation;

    public ButtonColorClickListener(Callback callback, int color) {
      this.callbackSoftReference = new SoftReference<Callback>(callback);
      this.color = color;
    }

    @Override
    public void onClick(View v) {
      Callback callback = callbackSoftReference.get();
      if (callback != null) {
        callback.colorChanged(color);
      }
      v.startAnimation(getAnimation());
    }

    private Animation getAnimation() {
      if (animation == null) {
        animation = new AlphaAnimation(1, 0);
        animation.setDuration(50);
        animation.setInterpolator(new LinearInterpolator());
      }
      return animation;
    }
  }

  /*@Override
  public void onViewCreated(View view, Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);

    ColorPickerView colorPickerView = (ColorPickerView) view.findViewById(R.id.color_picker);
    colorPickerView.setOnColorChangedListener(new ColorPickerDialog.OnColorChangedListener() {
      @Override
      public void colorChanged(int color) {
        Toast.makeText(getActivity(), "Clicked color", Toast.LENGTH_LONG).show();
      }
    });

  }*/
}

File: app/src/main/java/dk/eightyplus/annoto/fragment/ColorPaletteFragment.kt
package dk.eightyplus.annoto.fragment

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.view.animation.LinearInterpolator
import android.widget.ImageButton
import android.widget.LinearLayout
import dk.eightyplus.annoto.Callback
import dk.eightyplus.annoto.Keys
import dk.eightyplus.annoto.R
import dk.eightyplus.annoto.dialog.ColorPickerDialog
import dk.eightyplus.annoto.dialog.ColorPickerView
import java.lang.ref.SoftReference

/**
 * User: fries
 * Date: 15/07/14
 * Time: 14.00
 */
class ColorPaletteFragment : DialogFragment() {

    private var callbackSoftReference: SoftReference<Callback>? = null
    private var softPreferences: SoftReference<SharedPreferences>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.callbackSoftReference = SoftReference<Callback>(activity as Callback?)
        this.softPreferences = SoftReference<SharedPreferences>(activity?.getSharedPreferences(Keys.PREFERENCES, Context.MODE_PRIVATE))
    }

    fun preference(key: String, defaultValue: Int = -1): Int {
        val preferences = softPreferences?.get()
        return preferences?.getInt(key, defaultValue) ?: defaultValue
    }

    fun saveColorPreference(color: Int) {
        val preferences = softPreferences?.get() ?: return
        var lastColorSlot = preference(Keys.COLOR_LAST, -1)

        lastColorSlot++
        if (lastColorSlot > 5) {
            lastColorSlot = 0
        }

        val edit = preferences.edit()
        edit.putInt(Keys.COLOR + lastColorSlot, color)
        edit.putInt(Keys.COLOR_LAST, lastColorSlot)
        edit.apply()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.color_palette, container, false)

        val callback = callbackSoftReference?.get() ?: return view

        val linearLayoutTop = view.findViewById<View>(R.id.color_top) as LinearLayout
        val blackWhite = view.findViewById<View>(R.id.color_static1) as LinearLayout
        val grayScale = view.findViewById<View>(R.id.color_static2) as LinearLayout
        val rainbow = view.findViewById<View>(R.id.color_rainbow) as LinearLayout
        val linearLayoutDynamic = view.findViewById<View>(R.id.color_dynamic) as LinearLayout

        val bwColors = intArrayOf(R.color.black, R.color.white)
        for (color in bwColors) {
            val hexColor = callback.hexColorForResourceId(color)

            val button = inflater.inflate(R.layout.color_button, blackWhite, false) as ImageButton
            button.setBackgroundColor(hexColor)
            blackWhite.addView(button)

            button.setOnClickListener(ButtonColorClickListener(callback, hexColor))
        }
        val randomButton = inflater.inflate(R.layout.color_button, blackWhite, false) as ImageButton
        randomButton.setBackgroundResource(android.R.drawable.ic_menu_help)
        randomButton.setOnClickListener(ButtonColorClickListener(callback, 0))
        blackWhite.addView(randomButton)

        val grayColors = intArrayOf(R.color.dark_grey, R.color.gray, R.color.light_grey)
        for (color in grayColors) {
            val hexColor = callback.hexColorForResourceId(color)

            val button = inflater.inflate(R.layout.color_button, grayScale, false) as ImageButton
            button.setBackgroundColor(hexColor)
            grayScale.addView(button)

            button.setOnClickListener(ButtonColorClickListener(callback, hexColor))
        }

        val colorPickerView = ColorPickerView(inflater.context, object : ColorPickerDialog.OnColorChangedListener {
            private var lastColor = 0

            override fun colorChanged(color: Int) {
                callback.colorChanged(color)

                if (lastColor != color) {
                    lastColor = color
                    saveColorPreference(color)
                    updatePickedColors(linearLayoutDynamic, inflater, callback)
                    linearLayoutDynamic.invalidate()
                }
            }
        }, 0)
        linearLayoutTop.addView(colorPickerView)

        val colors = intArrayOf(R.color.red, R.color.Orange, R.color.yellow, R.color.green, R.color.blue, R.color.purple)
        for (color in colors) {
            val hexColor = callback.hexColorForResourceId(color)

            val button = inflater.inflate(R.layout.color_button, rainbow, false) as ImageButton
            button.setBackgroundColor(hexColor)
            rainbow.addView(button)

            button.setOnClickListener(ButtonColorClickListener(callback, hexColor))
        }

        updatePickedColors(linearLayoutDynamic, inflater, callback)

        return view
    }

    private fun updatePickedColors(linearLayoutDynamic: LinearLayout, inflater: LayoutInflater, callback: Callback) {
        linearLayoutDynamic.removeAllViews()
        val lastColorSlot = preference(Keys.COLOR_LAST, -1)
        if (lastColorSlot >= 0) {
            for (i in 0..5) {
                val hexWhiteColor = callback.hexColorForResourceId(R.color.white)
                val hexColor = preference(Keys.COLOR + i, hexWhiteColor)

                val button = inflater.inflate(R.layout.color_button, linearLayoutDynamic, false) as ImageButton
                button.setBackgroundColor(hexColor)
                linearLayoutDynamic.addView(button)

                button.setOnClickListener(ButtonColorClickListener(callback, hexColor))
            }
        }
    }


    private class ButtonColorClickListener(callback: Callback, private val color: Int) : View.OnClickListener {

        private val callbackSoftReference: SoftReference<Callback> = SoftReference(callback)
        private var animation: Animation? = null

        override fun onClick(v: View) {
            val callback = callbackSoftReference.get() ?: return
            callback.colorChanged(color)
            v.startAnimation(getAnimation())
        }

        private fun getAnimation(): Animation {
            var animation = animation
            if (animation == null) {
                animation = AlphaAnimation(1f, 0f)
                animation.duration = 50
                animation.interpolator = LinearInterpolator()
                this.animation = animation
            }
            return animation
        }
    }
}
